src/controller/launcher/webots_controller.c

```diff
@@ -0,0 +1,834 @@
+    fprintf(stderr, "Controller path '%s' is a directory. Please specify a path to an existing controller file.\n", controller);
+    exit(1);
+  };
+
+  // Compute path to controller file
+  char *controller_name = strrchr(controller, PATH_SEPARATOR[0]);
+  // Get extension from controller name (robust against relative paths)
+  if (!controller_name)
+    controller_extension = strrchr(controller, '.') == NULL ? NULL : strdup(strrchr(controller, '.'));
+  else
+    controller_extension = strrchr(controller_name, '.') == NULL ? NULL : strdup(strrchr(controller_name, '.'));
+#ifdef _WIN32
+  controller_path = strdup(".\\");
+#else
+  controller_path = strdup("./");
+#endif
+  if (strrchr(controller, PATH_SEPARATOR[0])) {
+    const size_t controller_file_size = strlen(strrchr(controller, PATH_SEPARATOR[0])) - 1;
+    const size_t controller_size = strlen(controller);
+    const size_t controller_path_size = controller_size - controller_file_size;
+    char *controller_path_tmp = malloc(controller_path_size + 1);
+    strncpy(controller_path_tmp, controller, controller_path_size);
+    controller_path_tmp[controller_path_size] = '\0';
+
+    // Change to controller directory and edit controller file path
+    chdir(controller_path_tmp);
+    const size_t new_controller_size = snprintf(NULL, 0, "%s%s", controller_path, controller_name + 1) + 1;
+    char *tmp_realloc = realloc(controller, new_controller_size);
+    if (!tmp_realloc)
+      exit(1);
+    controller = tmp_realloc;
+    snprintf(controller, new_controller_size, "%s%s", controller_path, controller_name + 1);
+  } else {
+    // Add current relative path to controller for execvp() function
+    const size_t controller_size = strlen(controller);
+    char *tmp_realloc = realloc(controller, controller_size + 3);
+    if (!tmp_realloc)
+      exit(1);
+    controller = tmp_realloc;
+    memmove(controller + 2, controller, controller_size + 1);
+    memcpy(controller, controller_path, 2);
+  }
+
+  // Parse eventual runtime.ini file
+  parse_runtime_ini();
+
+  // Executable controller
+  if (!controller_extension || strcmp(controller_extension, ".exe") == 0) {
+    exec_java_config_environment();
+#ifdef _WIN32
+    const char *const new_argv[] = {controller, NULL};
+    _spawnvpe(_P_WAIT, new_argv[0], new_argv, NULL);
+#else
+    char *new_argv[] = {controller, NULL};
+    execvp(new_argv[0], new_argv);
+#endif
+  }
+  // Python controller
+  else if (strcmp(controller_extension, ".py") == 0) {
+    python_config_environment();
+#ifdef _WIN32
+    char *controller_formated = NULL;
+    if (strstr(controller, " ") != NULL) {
+      const size_t controller_formated_size = snprintf(NULL, 0, "\"%s\"", controller) + 1;
+      controller_formated = malloc(controller_formated_size);
+      sprintf(controller_formated, "\"%s\"", controller);
+    } else
+      controller_formated = strdup(controller);
+
+    const char *const new_argv[] = {"python", controller_formated, NULL};
```

Comment:
If we use the `--stdout-redirect` option, we should add the "-u" option to the python command line so that the `stdout` is actually flushed (e.g., sent to Webots), otherwise it is buffered by default and doesn't show up in Webots console unless the user explicitly flushes `stdout`.

Reply:
Nice catch! Thank you!